### DISCUSSION TITLE
Queries around pull_request and workflow_run for PR testing

### DISCUSSION THREAD

**Discussion by Sandeepyadav93** [community/community#36794] [Code Security]
We are building CI for our repository where we are developing a containerized application. Whenever someone pushes a PR from their local fork to our organization's repo, We want to build container images using the PR content and push those container images to a Public registry(i.e GitHub registry/quay). We then want to test by running the application using container images(build from PR content) on our staging Kubernetes/Openshift cluster and upon successful CI pass - We want to merge the PR after manual reviews.

To push containers to registries and to connect to our Kubernetes cluster - We need credentials (stored in our GitHub organization/repository secrets). From my tests/reading doc, My understanding is pull_request doesn't have access to secrets, pull_target_request has secrets access but they run in the context of the target repository of the PR and using ${{ github.event.pull_request.head.sha }} is not recommended.[1]

A possible recommendation in [1] is to use two different workflows:-
 
A) In the first workflow use pull_request and build/push artifacts

```
name: Receive PR

# read-only repo token
# no access to secrets
on:
  pull_request:
```

B) Use workflow_run where we can pull artifacts from the jobs defined in the above workflow

```
name: Comment on the pull request

# read-write repo token
# access to secrets
on:
  workflow_run:
    workflows: ["Receive PR"]
    types:
      - completed
# Some logic to pull Artificats from jobs triggers from the above workflow
```

Question:

1) In the above example the second workflow just comments on PR, For our use case we don't want comments on PR, we want the CI job to report on pull_request. How can we use workflow_run to also report the job result to PR and not just comment?

2) Our artifacts from PR content are container images and not some files/output which be uploaded using actions/upload-artifact@v2. Because we want to upload container images to registries we need access to the secret.  How can we push container images to a public registry using pull_request/pull_request_target without compromising on security?

[1] https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

**Comment by ossanna16** [comment]
Thank you for asking a good question @Sandeepyadav93, I'm glad you found our community 🙂
